src/compiler/chainrules.jl

```diff
@@ -0,0 +1,31 @@
+# if there is no rrule defined this is the method that will be hit
+const chainrules_fallback_method = which(rrule, Tuple{Any})
+
+function has_chain_rrule(T)
+  m = meta(Tuple{typeof(rrule),T.parameters...})
+  if m.method === chainrules_fallback_method
+    return false, m.code.edges
+  else
+    return true, nothing
+  end
+end
+
+# For now we are just not going to deal with thunks
+wrap_chainrules(x) = conj(unthunk(x))
+wrap_chainrules(x::Tuple) = map(wrap_chainrules, x)
+
+function chain_rrule(f, args...)
+  #@info "Using ChainRule" f, typeof.(args)
+  y, back = rrule(f, args...)
+
+  zpullback(dy) = wrap_chainrules(back(dy))
+  # `nothing->nothing` can be deleted after https://github.com/FluxML/Zygote.jl/issues/603
+  # though it might be worth keeping as a performance optimization (benchmarking pending)
+  zpullback(::Nothing) = nothing
+
+  y, zpullback
+end
+
+# Don't need to conjugate these rrules, because of reasons I don't understand
+chain_rrule(f::typeof(abs), x::Number) = rrule(f, x)
+chain_rrule(f::typeof(abs2), x::Number) = rrule(f, x)
```

Comment:
This is concerning. I know it mirrors how we used DiffRules, but that ruleset wasn't user-extensible. In effect we're assuming all rules added to CR are analytic and will silently get incorrect results if that's not the case, which seems certain to cause problems down the line.

Reply:
Right now it is a requirement that all rules in ChainRules are analytic.

Reply:
Doesn't this imply you have to delete the rules for `abs` and `abs2` and define them in Zygote instead?

Reply:
probably  should ideally.
Right now happy mirroring DIffRules.
Definately thinking more on complex numbers has to happen still.
I am hoping @willtebbutt  is thinking about that right now

Reply:
I think they do need to be put in Zygote as hittind the fallback in Zygote to compute `abs(::Complex)` goes to bad places